### ISSUE TITLE
Rationale is incorrect

### ISSUE THREAD

**Issue by domenic** [thejameskyle/proposal-promise-access-internal-fields#2]
> You're using import() to load a component async. While the component is loading you want to show a loading state.
>
> However, if you have loaded the component already you run into a problem with import() being async. You can't wait until the next tick because your render is synchronous, and you don't want to show a flash of a loading frame for a single frame.

This misunderstands promises and rendering. Given

```js
document.body.innerHTML = "1";
Promise.resolve().then(() => {
  document.body.innerHTML = "2";
});
```

it is mandated by the spec that "1" is never painted to the screen. That is because microtasks take place before any tasks, including before animation frame tasks. Animation frame tasks only run every 16 milliseconds anyway, so you'd have to do something like

```js
document.body.innerHTML = "1";
delay(17).then(() => {
  document.body.innerHTML = "2";
});
```

to guarantee a one-frame flash. (You could still get a flash with a number between 0 and 17 if you were midway through a frame when this code ran, of course.)

As such this proposal is not very motivated. Already-fulfilled promises will *never* cause rendering flashes of the type discussed here.

(The same applies for fetch() BTW.)

**Comment by domenic** [comment]
Another issue is that it's a bad motivation to not want to show rendering flashes for import() of already-resolved modules, by testing their internal state, but decide that it's fine to show import() for modules that load in 50 milliseconds because the browser is quite fast (e.g. local development, memory caching, HTTP caching, etc.).

So this whole problem should be addressed by the library which is doing a bad job of creating loading screens, by e.g. not showing loading screens for 200 milliseconds or so. That is a much better cutoff than just arbitrarily deciding that already-fulfilled promises should not cause loading screens.

**Comment by domenic** [comment]
It's also worth noting that rAF is guaranteed to run before updating the rendering, so the following code:

```js
document.body.innerHTML = "1";
Promise.resolve().then(() => {
  requestAnimationFrame(() => {
    document.body.innerHTML = "2";
  });
});
```

is *also* guaranteed to never render "1".

**Comment by darrenscerri** [comment]
The reasoning might be that the `render` function is still invoked twice, once for when the promise is still pending and once when it is fulfilled, even though nothing is actually rendered on screen due to how rendering works. In situations such as server-side rendering or any other non-render situations, it might be important that the first function invocation when the promise is still pending does not occur.

**Comment by domenic** [comment]
That sounds like a deficiency in the framework then, not something that needs to be solved at the language level.

**Comment by darrenscerri** [comment]
Don't think this is something that a framework should support. If `import()` supports async module loading as a Promise, it might be ideal to have support for ensuring synchronous execution if the module is already loaded.

**Comment by domenic** [comment]
> it might be ideal to have support for ensuring synchronous execution if the module is already loaded.

The whole point of this thread is that a statement like this, or the one in the readme, is not justified. As shown here, there is no benefit of synchronous execution over microtask execution.

Maybe there are deficient frameworks which cannot deal with a single microtask of delay, but that's purely a framework problem: the platform has no limitations here that force a flash of unrendered content. The framework just isn't doing it's job.

**Comment by jamiebuilds** [comment]
Updated the example in the rationale to be more correct. Which is also now more realistic.

```js
function syncRender() {
  let component = null;
  let promiseComponent = import("./component").then(loadedComponent => {
    component = loadedComponent.default;
  });

  if (Promise.isResolved(promise)) {
    component = Promise.getValue(promise);
  }

  if (component !== null) {
    draw(component);
  } else {
    draw("Loading...");
  }

  requestAnimationFrame(syncRender);
}

requestAnimationFrame(syncRender);
```

This is a more realistic scenario because within a library/framework context you would want to defer `import()`-ing the module until you actually need it, which is often not discovered until you go to render it.

**Comment by jamiebuilds** [comment]
I do have a library that does exactly this ([react-loadable](https://github.com/thejameskyle/react-loadable)) and it does have this problem because I don't know I need to load anything until someone has attempted to render it.

> So this whole problem should be addressed by the library which is doing a bad job of creating loading screens, by e.g. not showing loading screens for 200 milliseconds or so. That is a much better cutoff than just arbitrarily deciding that already-fulfilled promises should not cause loading screens.

My library does do this, and yes it is still a problem that a blank screen flashes for a single frame.

**Comment by domenic** [comment]
Can you explain why this code does not work?

```js
let component = null;
let handle;
function syncRender() {
  if (component !== null) {
    draw(component);
  } else {
    draw("Loading...");
    import("./component").then(loadedComponent => {
      component = loadedComponent.default;
      cancelAnimationFrame(handle);
      requestAnimationFrame(syncRender);
    });
  }

  handle = requestAnimationFrame(syncRender);
}

requestAnimationFrame(syncRender);
```

Edit: I am 95% sure this works just as well with less complication, since rAF always occurs before painting so the next `syncRender()` call after setting component will see the update.

```js
let component = null;
function syncRender() {
  if (component !== null) {
    draw(component);
  } else {
    draw("Loading...");
    import("./component").then(loadedComponent => {
      component = loadedComponent.default;
    });
  }

  requestAnimationFrame(syncRender);
}

requestAnimationFrame(syncRender);
```

If `import("./component")` returns a fulfilled promise, then `"Loading..."` will never be drawn, in the same guarantee as in the OP.

**Comment by jamiebuilds** [comment]
The issue there is that it ignores how people bundle together modules and just because `component` is `null` does not mean that the module isn't already available.

Additionally, the eventual internal value of a promise isn't always going to be local like that. Ex: Libraries are meant to abstract away this kind of logic.

**Comment by domenic** [comment]
That isn't very satisfying, unfortunately. I've given you a straightforward change to your code that allows you to fix your library without introducing new language features. I have no choice but to oppose this proposal on the grounds that it is unmotivated.

We should not force the language to track state for you because a library is unwilling to make the miniscule code changes necessary to do so.

**Comment by jamiebuilds** [comment]
These are only miniscule code changes when you are writing the exact code above and when you ignore my previous statement. React has been rewriting it's entire engine to allow things like this to exist. It would be a significant change for many others.

**Comment by sebmarkbage** [comment]
```js
document.body.innerHTML = "1";
Promise.resolve().then(() => {
  document.body.innerHTML = "2";
});
```

is more expensive than:

```js
const value = Promise.getValue(Promise.resolve());
if (value) {
  document.body.innerHTML = "2";
} else {
  document.body.innerHTML = "1";
}
```

Once you have a stack of things that need to either be restored or tracked all the way down the chain it becomes even more expensive.

This example is demonstrating this as a micro-example but it doesn't take much to make the `innerHTML` string longer to demonstrate why this is a bottleneck. Then you scale this to deeper stacks and more invocations and the overhead is very significant.

**Comment by domenic** [comment]
@sebmarkbage it's pretty easy to work around that of course:

```js
let component;
maybeAlreadyFulfiled.then(v => {
  component = v;
  document.body.innerHTML = "2";
});

Promise.resolve().then(() => {
  if (!component) {
    document.body.innerHTML = "1";
  }
});
```

I'm going to disengage from this thread until we see a good-faith effort to create examples which are not trivially fixable. (Hint: no such example can have anything to do with events that occur every 16 ms, like painting.) Until then my opposition to the proposal stands.

**Comment by sebmarkbage** [comment]
In your example, you introduced `component` in a closure context shared by two callbacks and a noop Promise.

Typically it would be more that gets captured from the stack:

```js
let moreContent = compute(...);
let component;
maybeAlreadyFulfilled.then(v => {
  component = v;
  document.body.innerHTML = "2" + moreContent;
});

Promise.resolve().then(() => {
  if (!component) {
    document.body.innerHTML = "1";
  }
});
```

It may seem like trivial overhead in a simple example like this but when this scales up into a whole system it is not, and it very difficult to unwrap.

If `async/await` can be optimized to near zero-overhead compared to stacks then fine, we can probably convert our entire stacks and frameworks to `async/await` and then it might work (and Multicore OCaml has demonstrated that this can work well).

However, until then, we just can't use Promises all the way down and especially not unnecessary promises like in your solution.

If we can't read the internal state then we probably will just suggest that people use polyfills that allows for this, as an optimization.

**Comment by bterlson** [comment]
> I'm going to disengage from this thread until we see a good-faith effort to create examples which are not trivially fixable. Until then my opposition to the proposal stands.

This is uncalled for - people are simply trying to understand your objections. It is not bad faith to disagree or even to be wrong :-P

**Comment by jamiebuilds** [comment]
@domenic I'm gonna ask you to please be politer to me. I put this proposal up as a place to work out details of something I've been unable to solve in user land to see if JavaScript can better solve it. I have not brought this proposal to the committee at all and I did not ask for your feedback yet. This thread was an inappropriate way to communicate and I don't appreciate [this tweet](https://twitter.com/domenic/status/841862574610239489).

**Comment by matthewp** [comment]
Setting aside the issue of painting, is there any guarantee that a module that has already been imported could be retrieved synchronously using this API? Meaning, are we sure that the host environment doesn't need to perform some type of asynchronous operation under the hood to retrieve an already-imported module?  I ask because I do not know.
